src/browser/fake_network.js

```diff
@@ -961,125 +1042,222 @@ TCPConnection.prototype.accept = function(packet) {
  * @param {Uint8Array} data
  */
 TCPConnection.prototype.write = function(data) {
-    if(this.send_buffer.length > 0) {
-        // TODO: Pretty inefficient
-        let concat = new Uint8Array(this.send_buffer.byteLength + data.byteLength);
-        concat.set(this.send_buffer, 0);
-        concat.set(data, this.send_buffer.byteLength);
-        this.send_buffer = concat;
-    } else {
-        this.send_buffer = data;
+    if(!this.in_active_close) {
+        this.send_buffer.write(data);
     }
     this.pump();
 };
 
 TCPConnection.prototype.close = function() {
-    this.state = TCP_STATE_FIN_WAIT_1;
-    let reply = this.ipv4_reply();
-    reply.tcp.fin = true;
-    this.net.receive(make_packet(reply));
+    if(!this.in_active_close) {
+        this.in_active_close = true;
+        let next_state;
+        if(this.state === TCP_STATE_ESTABLISHED || this.state === TCP_STATE_SYN_RECEIVED) {
+            next_state = TCP_STATE_FIN_WAIT_1;
+        }
+        else if(this.state === TCP_STATE_CLOSE_WAIT) {
+            next_state = TCP_STATE_LAST_ACK;
+        }
+        else {
+            if(this.state !== TCP_STATE_SYN_SENT) {
+                dbg_log(`TCP[${this.tuple}]: active close in unexpected state "${this.state}"`, LOG_FETCH);
+            }
+            this.release();
+            return;
+        }
+
+        if(this.send_buffer.length || this.pending) {
+            // dbg_log(`TCP[${this.tuple}]: active close, delaying FIN in state "${this.state}", delayed next "${next_state}"`, LOG_FETCH);
+            this.delayed_send_fin = true;
+            this.delayed_state = next_state;
+        }
+        else {
+            // dbg_log(`TCP[${this.tuple}]: active close, sending FIN in state "${this.state}", next "${next_state}"`, LOG_FETCH);
+            this.state = next_state;
+            const reply = this.ipv4_reply();
+            reply.tcp.fin = true;
+            this.net.receive(make_packet(this.net.eth_encoder_buf, reply));
+        }
+    }
     this.pump();
 };
 
-TCPConnection.prototype.pump = function() {
-
-    if(this.send_buffer.length > 0 && !this.pending) {
-        let data = this.send_buffer.subarray(0, 500);
-        let reply = this.ipv4_reply();
+TCPConnection.prototype.release = function() {
+    if(this.net.tcp_conn[this.tuple]) {
+        // dbg_log(`TCP[${this.tuple}]: connection closed in state "${this.state}"`, LOG_FETCH);
+        this.state = TCP_STATE_CLOSED;
+        delete this.net.tcp_conn[this.tuple];
+    }
+};
 
-        this.pending = true;
-        if(this.send_buffer.length < 1) reply.tcp.fin = true;
+TCPConnection.prototype.pump = function() {
+    if(this.send_buffer.length && !this.pending) {
+        const data = this.send_chunk_buf;
+        const n_ready = this.send_buffer.peek(data);
+        const reply = this.ipv4_reply();
         reply.tcp.psh = true;
-        reply.tcp_data = data;
-        this.net.receive(make_packet(reply));
+        reply.tcp_data = data.subarray(0, n_ready);
```

Comment:
I believe this line is not necessary

Reply:
I'm not sure I understand. `reply.tcp_data` is needed, it's the TCP payload for the packet we create with `make_packet()` here. It is eventually used further down the call chain in `write_tcp()`.

Reply:
Sorry, I meant the `.subarray()` shouldn't be needed.

Reply:
Well, `data` has a fixed capacity of 1460 bytes of which only `n_ready` bytes are used at this point, which can be less than the capacity.

If we remove this `.subarray()` then `write_tcp()` would somehow need to know `n_ready` explicitely together with `data`. It would need an extra field `reply.tcp_data_length` next to `reply.tcp_data` or such.

As it is only called once per generated TCP packet I would just leave it like it is. Also, `write_tcp()` passes this subarray to `view_set_array()` which also treats it as an array with implicit size.

Reply:
You're absolutely right, I was somehow under the impression that `data` was returned from `peek` and so already had the right length.